Hubble: add possibility to export flows to container logs

siegmund-heiss-ich opened the pull request:
This is a modification of the static exporter for flow logs which allows setting the `hubble-export-file-path` to `stdout`. This triggers not the creation of rotating files but the direct output to logs.

The benefit is the possibility to ship flow logs to a SIEM for continuous analysis without the deployment of a separate daemonset like used here: https://github.com/cilium/hubble-otel

I am happy for any suggestions.

michi-covalent reviewed (approved):
lgtm 🚀 

- could you squash these 2 commits into 1 commit?
- curious how it shows up in kubectl logs. could you post a sample output?

chancez reviewed (changes requested):
Seems reasonable to me. Overall, the code changes look good. Just a few doc updates needed:

- Can you update the [flag help text](https://github.com/cilium/cilium/blob/main/daemon/cmd/daemon_main.go#L927). Make sure to re-run `make -C Documentation update-cmdref` after updating.
- Can you [update the usage documentation](https://github.com/cilium/cilium/blob/main/Documentation/observability/hubble-exporter.rst) as well?

qmonnet reviewed (approved):
The changes look good, although I have two generic remarks on the commit:

- Could you please add the context/motivation for the commit to the commit description? Basically what you currently have in the PR description, but many developers find it easier to get the information from the commit logs when running `git blame` rather than looking online for the corresponding pull request.
- Would you please consider using a valid email address in your sign-off? Although it's rare, we've had to reach out to contributors in the past, and emails help for that.

rolinh commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -140,5 +144,19 @@ func (e *exporter) OnDecodedEvent(_ context.Context, ev *v1.Event) (bool, error)
+			e.logger.WithError(err).Error("Failed to marshal event")
+			return false, err
+		}
+		e.logger.Info(string(jsonData))
```
I think it'd be much easier to post-process in SIEM and logs databases if we log the JSON flow as a separate field. It should help for use-cases where hubble flows need to be processed or indexed separately from the rest of the logs (ie JSON parsed or else).

```suggestion
		e.logger.WithField("flow", string(jsonData)).Info("Recorded Hubble flow")
```

siegmund-heiss-ich replied:
Thanks four your suggestion! I changed it accordingly.

glibsm commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -140,5 +144,19 @@ func (e *exporter) OnDecodedEvent(_ context.Context, ev *v1.Event) (bool, error)
 		return false, nil
 	}
-	return false, e.encoder.Encode(res)
+	// Check if we are logging directly
```
This functionality is probably better implemented as a simple writer that prints to stdout, rather than adding all this logic into the `OnDecodedEvent`

siegmund-heiss-ich replied:
That was exactly my first attempt. But I thought it wouldn't get merged if the logger isn't used. Also then we don't have additional fields. Sure, performance wise it would be better.

siegmund-heiss-ich replied:
With writing directly to stdout we also would get rid of the escaping chars, which makes the json string more useful.

siegmund-heiss-ich replied:
Changed it accordingly.

glibsm commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -140,5 +144,19 @@ func (e *exporter) OnDecodedEvent(_ context.Context, ev *v1.Event) (bool, error)
+			e.logger.WithError(err).Error("Failed to marshal event")
+			return false, err
+		}
+		e.logger.WithField("flow", string(jsonData)).Info("Recorded Hubble flow")
```
We may want to consider the performance cost of this going through logrus. There is a lot of copy involved for the `WithField` call and we may need a more efficient solution here.

siegmund-heiss-ich replied:
The usage of `WithField` was suggested by @rolinh 3 days ago.

glibsm reviewed (changes requested):
Requesting changes to halt the PR for a little bit so it's not merged accidentally.

glibsm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -202,7 +202,7 @@ cilium-agent [flags]
       --hubble-export-file-max-backups int                        Number of rotated Hubble export files to keep. (default 5)
       --hubble-export-file-max-size-mb int                        Size in MB at which to rotate Hubble export file. (default 10)
-      --hubble-export-file-path string                            Filepath to write Hubble events to.
+      --hubble-export-file-path stdout                            Filepath to write Hubble events to. By specifying stdout the flows are logged instead of written to a rotated file.
```
it's strange that the type of this has now changed from `string`.

chancez replied:
Yeah...that is quite weird actually.

siegmund-heiss-ich replied:
Is there anything I can/should do about this?
Btw. my commit is now rebased and all tests should work.

glibsm replied:
This file is auto-generated so if there is something we want to change about this, we'd have to track it down and fix the generator. There is no action on your part :)

glibsm reviewed (approved):
Sorry about the long delay, this PR slipped through my review queue somehow

siegmund-heiss-ich committed:
Add possibility to export flow logs to stdout

This is a modification of the static exporter for flow logs, which allows setting the hubble-export-file-path to stdout. This triggers not the creation of rotating files but the direct output to logs.
The benefit is the possibility to ship flow logs to a SIEM for continuous analysis without the deployment of a separate daemonset.

Signed-off-by: Alois Petutschnig <alois@petutschnig.net>
